File: src/server/crdt/generic/ormap_test.go
package crdt

import (
	"reflect"
	"sort"
	"testing"
)

// === Mock Dot Context Based CRDT for Testing ===

// type MockCRDT struct {
//     id        string
//     dotKernel *DotKernel[int]
// }

// func NewMockCRDT(id string) *MockCRDT {
//     return &MockCRDT{
//         id:        id,
//         dotKernel: NewDotKernel[int](),
//     }
// }

// func (m *MockCRDT) String() string {
//     return fmt.Sprintf("MockCRDT{id: %s, dotKernel: %v}", m.id, m.dotKernel)
// }

// func (m *MockCRDT) Context() *DotContext {
//     return m.dotKernel.Context()
// }

// func (m *MockCRDT) SetContext(ctx *DotContext) {
//     m.dotKernel.SetContext(ctx)
// }

// func (m *MockCRDT) Read() int {
//     value := 0
//     for _, v := range m.dotKernel.dotValues {
//         value = max(value, v)
//     }
//     return value
// }

// func (m *MockCRDT) Inc(amount int) *MockCRDT {
//     delta := NewMockCRDT(m.id)

//     for dot, v := range m.dotKernel.dotValues {
//         if dot.id == m.id {
//             newValue := v + amount

//             delta.dotKernel.Join(m.dotKernel.Add(m.id, newValue))

//             return delta
//         }
//     }

//     return delta
// }

// func (m *MockCRDT) Reset() *MockCRDT {
//     delta := NewMockCRDT(m.id)

//     delta.dotKernel = m.dotKernel.Reset()

//     return delta
// }

// func (m *MockCRDT) Join(other *MockCRDT) {
//     m.dotKernel.Join(other.dotKernel)
// }

// func (m *MockCRDT) NewEmpty(id string) *MockCRDT {
//     return NewMockCRDT(id)
// }

// func (m *MockCRDT) Clone() *MockCRDT {
//     clone := NewMockCRDT(m.id)
//     clone.dotKernel = m.dotKernel.Clone()
//     return clone
// }

// func (m *MockCRDT) IsNull() bool {
//     return m.Read() == 0
// }

// func (m *MockCRDT) Equal(other *MockCRDT) bool {
//     return m.Read() == other.Read() && reflect.DeepEqual(m.dotKernel, other.dotKernel)
// }

// === Test Auxiliary Functions ===

func equalORMaps(a, b *ORMap[string, *CCounter]) bool {
	if !reflect.DeepEqual(a.dotContext, b.dotContext) {
		return false
	}

	for key, valueA := range a.valueMap {
		valueB := b.Get(key)

		if !reflect.DeepEqual(valueA.dotKernel, valueB.dotKernel) {
			return false
		}
	}

	for key, valueB := range b.valueMap {
		valueA := a.Get(key)

		if !reflect.DeepEqual(valueA.dotKernel, valueB.dotKernel) {
			return false
		}
	}

	return true
}

// === ORMap Tests ===

func TestORMap_NewORMap(t *testing.T) {
	ormap := NewORMap[string]("replica1", NewCCounter)

	if ormap.replicaId != "replica1" {
		t.Errorf("Expected ORMap id to be 'replica1', got %s", ormap.replicaId)
	}
	if len(ormap.valueMap) != 0 {
		t.Errorf("Expected ORMap valueMap to be empty, got %v", ormap.valueMap)
	}
	if ormap.dotContext == nil {
		t.Errorf("Expected ORMap dotContext to be initialized, got nil")
	}
}

func TestORMap_Get(t *testing.T) {
	ormap := NewORMap[string]("replica1", NewCCounter)

	value1 := ormap.Get("key1")
	if !value1.IsNull() {
		t.Errorf("Expected created value for 'key1' to be nil: %v", value1)
	}

	value1.Inc(10)
	if ormap.Get("key1").Read() != 10 {
		t.Errorf("Expected value for 'key1' to be 10, got %d", ormap.Get("key1").Read())
	}
}

func TestORMap_Keys(t *testing.T) {
	ormap := NewORMap[string]("replica1", NewCCounter)
	ormap.Get("key1").Inc(10)
	ormap.Get("key2").Inc(20)

	keys := ormap.Keys()
	sort.Strings(keys)
	expectedKeys := []string{"key1", "key2"}

	if !reflect.DeepEqual(keys, expectedKeys) {
		t.Errorf("Expected keys %v, got %v", expectedKeys, keys)
	}
}

func TestORMap_KeysWithNullValues(t *testing.T) {
	ormap := NewORMap[string]("replica1", NewCCounter)
	ormap.Get("key1").Inc(5)
	ormap.Get("key2") // This will be null

	keys := ormap.Keys()
	expectedKeys := []string{"key1"}

	if !reflect.DeepEqual(keys, expectedKeys) {
		t.Errorf("Expected keys %v, got %v", expectedKeys, keys)
	}
}

func TestORMap_Apply(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap2 := ormap1.Clone()

	// Apply an increment operation
	delta := ormap1.Apply("key1", func(value *CCounter) *CCounter {
		return value.Inc(5)
	})

	// Check the updated value in the ORMap
	value := ormap1.Get("key1")
	if value.Read() != 5 {
		t.Errorf("Expected value for 'key1' to be 5, got %d", value.Read())
	}

	// Check the delta
	deltaValue := delta.Get("key1")
	if deltaValue.Read() != 5 {
		t.Errorf("Expected delta value for 'key1' to be 5, got %d", deltaValue.Read())
	}

	// Join delta to the cloned ORMap and verify equality
	ormap2.Join(delta)
	if !equalORMaps(ormap1, ormap2) {
		t.Logf("ORMap 1: %v", ormap1)
		t.Logf("ORMap 2: %v", ormap2)
		t.Errorf("Expected ormap1 and ormap2 to be equal after joining with delta")
	}
}

func TestORMap_Remove(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap2 := ormap1.Clone()

	delta := ormap1.Remove("key1")

	if len(ormap1.valueMap) != 0 {
		t.Errorf("Expected ORMap valueMap to be empty after removal, got %v", ormap1.valueMap)
	}

	ormap2.Join(delta)
	if !equalORMaps(ormap1, ormap2) {
		t.Errorf("Expected ormap1 and ormap2 to be equal after joining with delta")
	}
}

func TestORMap_Reset(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)
	ormap1.Get("key2").Inc(10)

	delta := ormap1.Reset()

	value1 := ormap1.Get("key1")
	value2 := ormap1.Get("key2")
	if value1.Read() != 0 || value2.Read() != 0 {
		t.Errorf("Expected all values in ORMap to be reset to 0")
	}

	deltaValue1 := delta.Get("key1")
	deltaValue2 := delta.Get("key2")
	if deltaValue1.Read() != 0 || deltaValue2.Read() != 0 {
		t.Errorf("Expected delta to have values resetted, got %v and %v", deltaValue1.Read(), deltaValue2.Read())
	}

	ormap2 := ormap1.Clone()
	ormap2.Join(delta)
	if !equalORMaps(ormap1, ormap2) {
		t.Logf("ORMap 1: %v", ormap1)
		t.Logf("ORMap 2: %v", ormap2)
		t.Errorf("Expected ormap1 and ormap2 to be equal after joining with delta")
	}
}

func TestORMap_Join(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := NewORMap[string]("replica2", NewCCounter)
	ormap2.Get("key1").Inc(10)
	ormap2.Get("key2").Inc(5)
	ormap1.Join(ormap2)

	if ormap1.Get("key1").Read() != 15 {
		t.Errorf("Expected ORMap to prefer higher value for 'key1', got %d", ormap1.Get("key1").Read())
	}
	if ormap1.Get("key2").Read() != 5 {
		t.Errorf("Expected ORMap to include 'key2' with value 5, got %v", ormap1.Get("key2").Read())
	}
}

func TestORMap_JoinWithEmptyORMap(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := NewORMap[string]("replica2", NewCCounter)

	ormap1.Join(ormap2)

	if len(ormap1.Keys()) != 1 {
		t.Errorf("Expected ORMap to still have 1 key after joining with empty ORMap, got %d", len(ormap1.Keys()))
	}
	if ormap1.Get("key1").Read() != 5 {
		t.Errorf("Expected ORMap to remain unchanged after joining with empty ORMap, got %d", ormap1.Get("key1").Read())
	}
}

func TestORMap_JoinIdempotent(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := ormap1.Clone()
	ormap1.Join(ormap2)

	if !equalORMaps(ormap1, ormap2) {
		t.Errorf("Expected ORMap to be unchanged after joining with itself")
	}
}

func TestORMap_JoinCommutative(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := NewORMap[string]("replica2", NewCCounter)
	ormap2.Get("key1").Inc(10)
	ormap2.Get("key2").Inc(5)

	// Join in one order
	joined1 := ormap1.Clone()
	joined1.Join(ormap2)

	// Join in the opposite order
	joined2 := ormap2.Clone()
	joined2.Join(ormap1)

	if !equalORMaps(joined1, joined2) {
		t.Logf("Joined 1: %v", joined1)
		t.Logf("Joined 2: %v", joined2)
		t.Errorf("Expected ORMap joins to be commutative, but got different results")
	}
}

func TestORMap_JoinIndependence(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := NewORMap[string]("replica2", NewCCounter)
	ormap2.Get("key2").Inc(10)

	// Join ormap2 into ormap1
	joined1 := ormap1.Clone()
	joined1.Join(ormap2)

	// Modify ormap2 after the join
	ormap2.Get("key2").Inc(10)

	if joined1.Get("key2").Read() != 10 {
		t.Errorf("Expected joined ORMap to remain unchanged after modifying the other ORMap, got %d", joined1.Get("key2").Read())
	}
}

func TestORMap_JoinAfterRemove(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := ormap1.Clone()
	delta := ormap1.Remove("key1")

	// t.Logf("Before merge: %v", ormap2)
	ormap2.Join(delta)

	if !equalORMaps(ormap1, ormap2) {
		t.Logf("ORMap 1: %v", ormap1)
		t.Logf("ORMap 2: %v", ormap2)
		t.Logf("Delta: %v", delta)
		t.Errorf("Expected ORMaps to be equal after joining with removal delta")
	}
}

func TestORMap_Clone(t *testing.T) {
	ormap1 := NewORMap[string]("replica1", NewCCounter)
	ormap1.Get("key1").Inc(5)

	ormap2 := ormap1.Clone()

	if !equalORMaps(ormap1, ormap2) {
		t.Errorf("Expected cloned ORMap to be equal to the original")
	}

	// Modify the clone and ensure original is unchanged
	ormap2.Get("key1").Inc(10)
	if ormap1.Get("key1").Read() != 5 {
		t.Errorf("Expected original ORMap to remain unchanged after modifying clone, got %d", ormap1.Get("key1").Read())
	}
}

func TestORMap_SetContext(t *testing.T) {
	ormap := NewORMap[string]("replica1", NewCCounter)
	ormap.Get("key1").Inc(5)

	newContext := NewDotContext()
	ormap.SetContext(newContext)

	if ormap.Context() != newContext {
		t.Errorf("Expected ORMap context to be updated, but it was not")
	}
	value := ormap.Get("key1")
	if value.Context() != newContext {
		t.Errorf("Expected value context to be updated, but it was not")
	}
}

File: src/client/types/crdt/generic/ormap.test.ts
import ORMap from './ormap';
import CCounter from './ccounter';
import DotContext from './dot-context';

function equalORMaps<K>(a: ORMap<K, CCounter>, b: ORMap<K, CCounter>): boolean {
    if (JSON.stringify(a.getContext()) !== JSON.stringify(b.getContext())) {
        return false;
    }

    const aKeys = Array.from(a.keys());
    const bKeys = Array.from(b.keys());

    if (aKeys.length !== bKeys.length) {
        return false;
    }

    for (const key of aKeys) {
        const valueA = a.get(key);
        const valueB = b.get(key);

        if (JSON.stringify(valueA['dotKernel']) !== JSON.stringify(valueB['dotKernel'])) {
            return false;
        }
    }

    return true;
}

describe('ORMap', () => {
    it('should create a new ORMap', () => {
        const ormap = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));

        expect(ormap['replicaId']).toBe('replica1');
        expect(ormap['valueMap'].size).toBe(0);
        expect(ormap.getContext()).not.toBeNull();
    });

    it('should get a value from the ORMap', () => {
        const ormap = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));

        const value1 = ormap.get('key1');
        expect(value1.isNull()).toBe(true);

        value1.inc(10);
        expect(ormap.get('key1').read()).toBe(10);
    });

    it('should return the keys of the ORMap', () => {
        const ormap = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap.get('key1').inc(10);
        ormap.get('key2').inc(20);

        const keys = Array.from(ormap.keys()).sort();
        expect(keys).toEqual(['key1', 'key2']);
    });

    it('should return keys excluding null values', () => {
        const ormap = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap.get('key1').inc(5);
        ormap.get('key2'); // This will be null

        const keys = Array.from(ormap.keys());
        expect(keys).toEqual(['key1']);
    });

    it('should apply an operation to a value in the ORMap', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        const ormap2 = ormap1.clone();

        const delta = ormap1.apply('key1', (value) => value.inc(5));

        const value = ormap1.get('key1');
        expect(value.read()).toBe(5);

        const deltaValue = delta.get('key1');
        expect(deltaValue.read()).toBe(5);

        ormap2.join(delta);
        expect(equalORMaps(ormap1, ormap2)).toBe(true);
    });

    it('should remove a value from the ORMap', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        const ormap2 = ormap1.clone();

        const delta = ormap1.remove('key1');

        expect(ormap1['valueMap'].size).toBe(0);

        ormap2.join(delta);
        expect(equalORMaps(ormap1, ormap2)).toBe(true);
    });

    it('should reset the ORMap', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);
        ormap1.get('key2').inc(10);

        const delta = ormap1.reset();

        expect(ormap1.get('key1').read()).toBe(0);
        expect(ormap1.get('key2').read()).toBe(0);

        const deltaValue1 = delta.get('key1');
        const deltaValue2 = delta.get('key2');
        expect(deltaValue1.read()).toBe(0);
        expect(deltaValue2.read()).toBe(0);

        const ormap2 = ormap1.clone();
        ormap2.join(delta);
        expect(equalORMaps(ormap1, ormap2)).toBe(true);
    });

    it('should join two ORMaps', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = new ORMap<string, CCounter>('replica2', (replicaId) => new CCounter(replicaId));
        ormap2.get('key1').inc(10);
        ormap2.get('key2').inc(5);

        ormap1.join(ormap2);

        expect(ormap1.get('key1').read()).toBe(15);
        expect(ormap1.get('key2').read()).toBe(5);
    });

    it('should join with an empty ORMap', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = new ORMap<string, CCounter>('replica2', (replicaId) => new CCounter(replicaId));

        ormap1.join(ormap2);

        expect(Array.from(ormap1.keys()).length).toBe(1);
        expect(ormap1.get('key1').read()).toBe(5);
    });

    it('should be idempotent when joining with itself', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = ormap1.clone();
        ormap1.join(ormap2);

        expect(equalORMaps(ormap1, ormap2)).toBe(true);
    });

    it('should be commutative when joining', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = new ORMap<string, CCounter>('replica2', (replicaId) => new CCounter(replicaId));
        ormap2.get('key1').inc(10);
        ormap2.get('key2').inc(5);

        const joined1 = ormap1.clone();
        joined1.join(ormap2);

        const joined2 = ormap2.clone();
        joined2.join(ormap1);

        expect(equalORMaps(joined1, joined2)).toBe(true);
    });

    it('should maintain independence after joining', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = new ORMap<string, CCounter>('replica2', (replicaId) => new CCounter(replicaId));
        ormap2.get('key2').inc(10);

        const joined1 = ormap1.clone();
        joined1.join(ormap2);

        ormap2.get('key2').inc(10);

        expect(joined1.get('key2').read()).toBe(10);
    });

    it('should handle joining after a removal', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = ormap1.clone();
        const delta = ormap1.remove('key1');

        ormap2.join(delta);

        expect(equalORMaps(ormap1, ormap2)).toBe(true);
    });

    it('should clone the ORMap', () => {
        const ormap1 = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap1.get('key1').inc(5);

        const ormap2 = ormap1.clone();

        expect(equalORMaps(ormap1, ormap2)).toBe(true);

        ormap2.get('key1').inc(10);
        expect(ormap1.get('key1').read()).toBe(5);
    });

    it('should set a new context', () => {
        const ormap = new ORMap<string, CCounter>('replica1', (replicaId) => new CCounter(replicaId));
        ormap.get('key1').inc(5);

        const newContext = new DotContext();
        ormap.setContext(newContext);

        expect(ormap.getContext()).toBe(newContext);
        expect(ormap.get('key1').getContext()).toBe(newContext);
    });
});